### PR TITLE
PYTHON-3472 - Add log messages to SDAM spec

### PR DESCRIPTION
**Pull request by NoahStapp**

### CODE REVIEW

File: pymongo/logger.py [hunk 1]
```diff
@@ -129,7 +141,9 @@ def _is_sensitive(self, doc_name: str) -> bool:
         )
 
         is_sensitive_hello = (
-            self._kwargs["commandName"] in _HELLO_COMMANDS and is_speculative_authenticate
+            "commandName" in self._kwargs
+            and self._kwargs["commandName"] in _HELLO_COMMANDS
+            and is_speculative_authenticate
```
Comment (Jibola) [inline pymongo/logger.py:146]:
```suggestion
            self._kwargs.get("commandName", None) in _HELLO_COMMANDS
            and is_speculative_authenticate
```

File: test/discovery_and_monitoring/rs/compatible.json [hunk 2]
```diff
@@ -16,7 +16,7 @@
               "b:27017"
             ],
             "minWireVersion": 0,
-            "maxWireVersion": 21
+            "maxWireVersion": 6
```
Comment (Jibola) [inline test/discovery_and_monitoring/rs/compatible.json:19]: Guessing this is just a spec requirement change?

File: test/unified_format.py [hunk 3]
```diff
@@ -1804,14 +1818,28 @@ def check_events(self, spec):
             else:
                 assert server_connection_id is None
 
+    def process_ignore_messages(self, ignore_logs, actual_logs):
+        final_logs = []
+        for log in actual_logs:
```
Comment (Jibola) [inline test/unified_format.py:1823]: NIT: Thoughts on this potential optimization?


actual_log_messages = {actual_log["data"]["message"] for actual_log in actual_logs}
return list(filter(lambda log: log["data"]["message"] not in actual_log_messages))

File: test/unified_format.py [hunk 4]
```diff
@@ -1804,14 +1818,28 @@ def check_events(self, spec):
...
+
     def check_log_messages(self, operations, spec):
         def format_logs(log_list):
             client_to_log = defaultdict(list)
             for log in log_list:
                 if log.module == "ocsp_support":
                     continue
                 data = json_util.loads(log.message)
-                client = data.pop("clientId")
+                client = data.pop("clientId") if "clientId" in data else data.pop("topologyId")
```
Comment (Jibola) [inline test/unified_format.py:1842]:
```suggestion
                client = data.pop("clientId", data.pop("topologyId"))
```

**Review by Jibola — DISMISSED** [review]
Overall LGTM!

Left non-blocking comments.

File: test/discovery_and_monitoring/rs/compatible.json (hunk 2 above)
Comment (NoahStapp) [reply test/discovery_and_monitoring/rs/compatible.json:19]: Yes.

File: test/unified_format.py (hunk 4 above)
Comment (NoahStapp) [reply test/unified_format.py:1842]: This is slick, nice.

File: test/unified_format.py (hunk 3 above)
Comment (NoahStapp) [reply test/unified_format.py:1823]: I like this optimization but it does deviate from the spec's statement that ["Matching rules used to match messages in `ignoreMessages` are identical to match rules used for `messages` matching".](https://github.com/mongodb/specifications/blob/a35de873621a244b88cad01aa3cd7c8da31a5ecd/source/unified-test-format/unified-test-format.md?plain=1#L1235) Is this a case where we need to stick closely to the spec, or can we do this much simpler way of matching?

File: test/unified_format.py (hunk 4 above)
Comment (NoahStapp) [reply test/unified_format.py]: Sadly doesn't work though, it tries to pop the default every time and causes a `KeyError`.

File: pymongo/asynchronous/monitor.py [hunk 5]
```diff
@@ -274,22 +287,31 @@ async def _check_once(self) -> ServerDescription:
...
         address = self._server_description.address
+        sd = self._server_description
+        awaited = bool(
+            self._pool.conns and self._stream and sd.is_server_type_known and sd.topology_version
+        )
         if self._publish:
             assert self._listeners is not None
-            sd = self._server_description
             # XXX: "awaited" could be incorrectly set to True in the rare case
             # the pool checkout closes and recreates a connection.
```
Comment (ShaneHarvey) [inline pymongo/asynchronous/monitor.py:297]: This comment should move along with the `awaited` var.

File: pymongo/synchronous/monitor.py [hunk 6]
```diff
@@ -302,6 +324,19 @@ def _check_once(self) -> ServerDescription:
...
+            if _SDAM_LOGGER.isEnabledFor(logging.DEBUG):
+                _debug_log(
+                    _SDAM_LOGGER,
+                    topologyId=self._topology._topology_id,
+                    driverConnectionId=conn.id,
+                    serverConnectionId=conn.server_connection_id,
+                    serverHost=address[0],
+                    serverPort=address[1],
+                    awaited=awaited,
+                    durationMS=round_trip_time,
```
Comment (ShaneHarvey) [inline pymongo/synchronous/monitor.py:336]: Here and elsewhere: round_trip_time is a float in seconds, but durationMS should be in milliseconds.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: pymongo/synchronous/monitor.py (hunk 6 above)
Comment (ShaneHarvey) [reply pymongo/synchronous/monitor.py]: Still need to make this change in the other locations where durationMS is passed.

File: pymongo/asynchronous/monitor.py [hunk 7]
```diff
@@ -257,10 +259,21 @@ async def _check_server(self) -> ServerDescription:
...
                 assert self._listeners is not None
                 self._listeners.publish_server_heartbeat_failed(address, duration, error, awaited)
+            if _SDAM_LOGGER.isEnabledFor(logging.DEBUG):
+                _debug_log(
+                    _SDAM_LOGGER,
+                    topologyId=self._topology._topology_id,
+                    serverHost=address[0],
+                    serverPort=address[1],
+                    awaited=awaited,
+                    durationMS=duration,
```
Comment (ShaneHarvey) [inline pymongo/asynchronous/monitor.py:273]: durationMS needs to be in ms here too.

File: pymongo/asynchronous/monitor.py [hunk 8]
```diff
@@ -274,22 +287,32 @@ async def _check_once(self) -> ServerDescription:
...
-                and self._stream
-                and sd.is_server_type_known
-                and sd.topology_version
-            )
             self._listeners.publish_server_heartbeat_started(address, awaited)
 
         if self._cancel_context and self._cancel_context.cancelled:
             await self._reset_connection()
         async with self._pool.checkout() as conn:
+            if _SDAM_LOGGER.isEnabledFor(logging.DEBUG):
```
Comment (ShaneHarvey) [inline pymongo/asynchronous/monitor.py:304]: Shouldn't this happen at the same point where publish_server_heartbeat_started is called?

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: pymongo/asynchronous/monitor.py (hunk 8 above)
Comment (NoahStapp) [reply pymongo/asynchronous/monitor.py:304]: In theory yes. However, we need a connection for the required logging information, and checking out the connection before publishing the heartbeat has a lot of side effects for other tests.

Comment (ShaneHarvey) [reply pymongo/asynchronous/monitor.py:304]: I would expect driverConnectionId to be None on the initial started event. Does the spec not expect that?

Comment (NoahStapp) [reply pymongo/asynchronous/monitor.py:304]: No such allowance is made: https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring-logging-and-monitoring.md#server-heartbeat-started-log-message.

File: test/unified_format.py (hunk 3 above)
Comment (Jibola) [reply test/unified_format.py:1823]: Good point. I'll say let's not then. 

The `self.match_evaluator.match_result()` function could be added in, but we lose readability so we'll keep it as is.

**Review by Jibola — APPROVED** [review]

File: pymongo/asynchronous/monitor.py (hunk 8 above)
Comment (ShaneHarvey) [reply pymongo/asynchronous/monitor.py:304]: Hmm this was intended to be fixed in DRIVERS-2677.

**Comment by ShaneHarvey** [comment]
Is this pypy failure related?:
```
=================================== FAILURES ===================================
_ TestConnectionLoggingConnectionLogging.test_Create_a_client,_run_a_command,_and_close_the_client _

self = <test.test_connection_logging.TestConnectionLoggingConnectionLogging testMethod=test_Create_a_client,_run_a_command,_and_close_the_client>

    def test_case(self):
>       self.run_scenario(spec)

test/unified_format.py:2069: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/unified_format.py:2013: in run_scenario
    self._run_scenario(spec, uri)
test/unified_format.py:2043: in _run_scenario
    self.check_log_messages(spec["operations"], expect_log_messages)
test/unified_format.py:1939: in check_log_messages
    formatted_logs = format_logs(cm.records)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

log_list = [<LogRecord: pymongo.topology, 10, /home/runner/work/mongo-python-driver/mongo-python-driver/pymongo/logger.py, 100, "...66cddd96f2b922fd71bcd93c"}, "message": "Connection pool ready", "serverHost": "localhost", "serverPort": 27017}">, ...]

    def format_logs(log_list):
        client_to_log = defaultdict(list)
        for log in log_list:
            if log.module == "ocsp_support":
                continue
>           data = json_util.loads(log.message)
E           AttributeError: 'LogRecord' object has no attribute 'message'

test/unified_format.py:1926: AttributeError
------------------------------ Captured log call -------------------------------
DEBUG    pymongo.topology:logger.py:100 {"topologyId": {"$oid": "66cddd96f2b922fd71bcd93c"}, "driverConnectionId": 1, "serverConnectionId": 901, "serverHost": "localhost", "serverPort": 27017, "awaited": true, "message": "Server heartbeat started"}
```

**Comment by NoahStapp** [comment]
> Is this pypy failure related?:
> 
> ```
> =================================== FAILURES ===================================
> _ TestConnectionLoggingConnectionLogging.test_Create_a_client,_run_a_command,_and_close_the_client _
> 
> self = <test.test_connection_logging.TestConnectionLoggingConnectionLogging testMethod=test_Create_a_client,_run_a_command,_and_close_the_client>
> 
>     def test_case(self):
> >       self.run_scenario(spec)
> 
> test/unified_format.py:2069: 
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> test/unified_format.py:2013: in run_scenario
>     self._run_scenario(spec, uri)
> test/unified_format.py:2043: in _run_scenario
>     self.check_log_messages(spec["operations"], expect_log_messages)
> test/unified_format.py:1939: in check_log_messages
>     formatted_logs = format_logs(cm.records)
> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
> 
> log_list = [<LogRecord: pymongo.topology, 10, /home/runner/work/mongo-python-driver/mongo-python-driver/pymongo/logger.py, 100, "...66cddd96f2b922fd71bcd93c"}, "message": "Connection pool ready", "serverHost": "localhost", "serverPort": 27017}">, ...]
> 
>     def format_logs(log_list):
>         client_to_log = defaultdict(list)
>         for log in log_list:
>             if log.module == "ocsp_support":
>                 continue
> >           data = json_util.loads(log.message)
> E           AttributeError: 'LogRecord' object has no attribute 'message'
> 
> test/unified_format.py:1926: AttributeError
> ------------------------------ Captured log call -------------------------------
> DEBUG    pymongo.topology:logger.py:100 {"topologyId": {"$oid": "66cddd96f2b922fd71bcd93c"}, "driverConnectionId": 1, "serverConnectionId": 901, "serverHost": "localhost", "serverPort": 27017, "awaited": true, "message": "Server heartbeat started"}
> ```

I wouldn't expect it to be related, but it doesn't appear to be failing on other PRs. According to the [logging docs](https://docs.python.org/2/library/logging.html#logrecord-attributes), `LogRecord.message` is set only when the formatter is called. We've seen this exact error before intermittently, but haven't tracked down the underlying cause.

**Pull request merged**